aten/src/ATen/native/TensorShape.cpp

```diff
@@ -2509,6 +2510,151 @@ Tensor & transpose_(Tensor & self, int64_t dim0, int64_t dim1) {
   return self;
 }
 
+namespace {
+// Transpose implementation for sparse compressed layouts
+// NB: We assume that dim1,dim0 have already been wrapped
+static inline Tensor sparse_compressed_transpose(
+    const Tensor& self,
+    int64_t dim0,
+    int64_t dim1) {
+  auto compressed_inds = AT_DISPATCH_ROW_SPARSE_COMPRESSED_LAYOUTS(
+      self.layout(),
+      "compressed_inds",
+      [&self]() { return self.crow_indices(); },
+      [&self]() { return self.ccol_indices(); });
+
+  auto plain_inds = AT_DISPATCH_ROW_SPARSE_COMPRESSED_LAYOUTS(
+      self.layout(),
+      "plain_inds",
+      [&self]() { return self.col_indices(); },
+      [&self]() { return self.row_indices(); });
+
+  const auto n_batch_dim = compressed_inds.dim() - 1;
+  const auto n_dense_dim = AT_DISPATCH_PLAIN_SPARSE_COMPRESSED_LAYOUTS(
+      self.layout(),
+      "n_dense_dim",
+      [&]() { return self.values().dim() - n_batch_dim - 1; },
+      [&]() { return self.values().dim() - n_batch_dim - 3; });
```

Comment:
Sorry, am I confusing something? For the CSR/CSR shouldn't it be either `self.dim() - n_batch_dim -1` or `self.values().dim() - 1`? Same for the other case.

Reply:
Sorry, I got confused, forgot about batchdims in values as well :)

Reply:
You bring up a good point though. The calculation of `n_dense_dim` can just use `self.dim()` and `n_batch_dim` and then does not need to dispatch based on unblocked/blocked layouts.